### PR TITLE
Improve xAxis ticks, thinner bottom margin

### LINKED ISSUE #4751: [P00 - Dashboard] Chart whitespace customization
There is too much whitespace on the line charts - this is mainly driven by how the X-axis diagonally shows values (see image)

Currently, only able to shorten the vertical space of the x-axis which really only cuts off the text (also not helpful)

Options:

(1) Allow for text wrapping, vertical, horizontal (as well as diagonal) placing of x-axis text

(2) Allow for option to not have values in x-axis to show up, relying completely on tooltip

Make sure these boxes are checked before submitting your issue - thank you!

- [ ] I have checked the superset logs for python stacktraces and included it here as text if any
- [ ] I have reproduced the issue with at least the latest released version of superset
- [ ] I have checked the issue tracker for the same issue and I haven't found one similar

JIRA - DPTOOLS 448

### PR DESCRIPTION
**Pull request by mistercrunch**
Introducing a new control `X Tick Layout` with options `['auto', 'flat', '45°', 'staggered']` for nvd3 visualizations.

nvd3 visualizations now do a much better job at minimizing the bottom margin in `auto` mode, and give more control to the user.

Also refactored quite a bit and aligned the controls across charts.

closes https://github.com/apache/incubator-superset/issues/4751

margins

### CODE REVIEW

**Review by betodealmeida — APPROVED** [review]
LGTM, but it seems like `superset/assets/visualizations/nvd3_vis.js` is becoming (already is?) unmanageable... should we split each nvd3 chart into a separate file?

**Comment by mistercrunch** [comment]
I agree on the [un] manageability of this file. The challenge is that the different visualizations both from a controls and API standpoint have a lot in common. A lot of code is shared in odd intricate ways. It's a though puzzle.

A balanced way to evolve this would be to refactor out portion of the code into smaller functions an unit test those. 

We've been talking about moving to `data-ui` / vx for a while, and maybe that's the end game and justify not spending too many cycles on `nvd3`.

**Comment by mistercrunch** [comment]
@graceguo-supercat @michellethomas would love for someone at Airbnb to sign off on this as it will affects a million line charts over there, but that should be for the best.

**Comment by williaster** [comment]
great addition. my only suggestion at this point is setting `flat` as the default because `staggered` isn't as polished / adds visual noise to the chart. thoughts @elibrumbaugh? 

as discussed offline, in the future it'd be useful to add a "concise"/"compact" version of the auto time format to help fix this problem. this could be used for axes and the "detailed" version could be used in tooltips.

**Comment by mistercrunch** [comment]
I'd also prefer `flat` to staggered as a default, but with the current default time formatter the labels overlap in some situation, depending on the time grain (some are wider) and on the width (d3.axis fits more or less tick density in ways we don't control).

Maybe next PR on this we'd go:
* new default `auto` format that's shorter than current `Adaptive formatting`
* use `flat` as new default
* somehow squeeze `Adaptive formatting` as the tooltip formatter (should be possible)

The magic formatting is the part where the devil is in the detail and people may not agree on the exact layout. Here's a link to the code we have now:
https://github.com/apache/incubator-superset/blob/master/superset/assets/javascripts/modules/dates.js#L15

Oh and we haven't even talked about locales :(

**Comment by williaster** [comment]
is there a rush to merge this? would it be worthwhile to add the shorter date formatting in this PR and set the default to `flat`?

the main concern I have is with saved slices, by merging this `staggered` is set for any slices saved for the foreseeable future and I'd guess users wouldn't go to the trouble to update them after the new concise time is added. another option would be a migration script to change them to `flat`, but that seems tricky?

**Comment by mistercrunch** [comment]
Actually the default setting is `auto`, which turns into staggered for time series. We can change what `auto` means in the future. Same goes with `Adaptive formatting`.

**Comment by mistercrunch** [comment]
But yeah, while this doesn't fix everything, it addresses many visualizations defects and bugs with the xAxis and makes much better use of real estate. It also does not change the content of the labels which could be controversial and could be considered breaking backwards compatibility.

It also improves the organization of axis-related controls for many chart types, making more consistent.

**Comment by williaster** [comment]
ah, that's nice about the `auto` setting. I'd like to get the `flat` in asap so users see staggered for the minimum amount of time, I'll begin work on the concise date logic!

**Comment by mistercrunch** [comment]
@williaster  I agree with `flat` but didn't want to drive it myself as it may upset some users since we have to redefine/shorten `Adaptive formatting` as a prerequisite  (I think the current definition was fine tuned by data scientists at Airbnb, forgot who but it's in the git history). It will also change the way thousands of line charts look like at Airbnb...

You have my blessing to figure out better defaults that will please a majority of users. One key point is that day-of-week is very important to both Airbnb and Lyft given the strong weekly cycles, I wouldn't kick this out, better removing `YYYY` than removing 3-letters day-of-week.

**Comment by mistercrunch** [comment]
I'll wait on @graceguo-supercat's review before merging this

File: superset/assets/javascripts/utils/common.js
Comment (GabeLoins) [inline superset/assets/javascripts/utils/common.js]: acknowledging I have very little context here- this seems like a bad road to go down. We use YAML for plenty of stuff over here and there have been many issues caused by someone not expecting `True`/`t`/`yes` to be true...

Comment (mistercrunch) [reply superset/assets/javascripts/utils/common.js]: It has to do with backward compatibility here. There's been a few instances of `CheckboxControl`  becoming a `SelectControl`. In this case the `show_brush` used to be checkbox and now is `auto`, `Yes` or 'No'.

Since it's not the first time and won't be the last, we decided to go this route. 

The alternatives would be to either:
* write database migrations to alter all slices and replace `True` with `'Yes'` but that's more work and db migration have maintenance overhead.
* systematically use `if (fd.show_brush === true || fd.show_brush === 'yes')` everywhere

**Comment by graceguo-supercat** [comment]
@mistercrunch This PR works good in airbnb data! please merge whenever you feel comfortable. Thanks!

**Pull request merged**
